teps/0130-pipeline-level-docker-daemon.md

```diff
@@ -0,0 +1,409 @@
+
+#### Related features in other CI/CD systems
+
+A few other CI/CD systems have sidecar-like features.
+
+The closest feature to Pipeline-level sidecars is Argo Workflows [daemon containers](https://argoproj.github.io/argo-workflows/walk-through/daemon-containers/) feature.
+Daemon containers share the same lifespan as a Template, which can be analogous to a Task or a Pipeline. (Argo Workflows also has a [sidecar feature](https://argoproj.github.io/argo-workflows/walk-through/sidecars/) similar to Tekton's Task sidecars.)
+
+Other CI/CD systems have features that are more analogous to Task sidecars in Tekton, and designed with integration tests in mind:
+* GitHub Actions [service containers](https://docs.github.com/en/actions/using-containerized-services/about-service-containers)
+* GitLab [services](https://docs.gitlab.com/ee/ci/services/)
+
+Both [GitLab](https://docs.gitlab.com/ee/ci/docker/using_docker_build.html#use-the-docker-executor-with-docker-in-docker)
+and [CircleCI](https://circleci.com/docs/executor-intro/#docker) provide Docker "executors" that can be used for running Docker builds.
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable. This may include API specs (though not always
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+
+If it's helpful to include workflow diagrams or any other related images,
+add them under "/teps/images/". It's upto the TEP author to choose the name
+of the file, but general guidance is to include at least TEP number in the
+file name, for example, "/teps/images/NNNN-workflow.jpg".
+-->
+
+This needs to be fleshed out in more detail, but the sidecar would likely be implemented as a deployment with one pod plus a service.
+If the sidecar has readiness or startup probes, the PipelineRun would wait for them to succeed before starting the Pipeline Tasks.
+
+## Design Evaluation
+
+### Reusability
+
+A few sidecar configurations, such as docker daemons and database containers, likely make up the majority of use cases.
+A major open question of this proposal is whether we want to introduce a concept of a reusable sidecar.
+For example, we could include a docker daemon sidecar in the catalog,
+and allow Pipelines to contain inline sidecar specifications or references to sidecars
+(stored on the hub, on the cluster, in git, etc, just like Tasks).
+Tasks would then need to document that they need a sidecar to work correctly (likely by accepting the sidecar address as a required parameter).
+
+### Simplicity
+
+The current user experience isn't flexible enough for many common docker-related use cases involving more complex logic than a build and push.
+
+### Flexibility
+
+Allowing users to specify Pipeline-level sidecars is more flexible than first class support for building images,
+like Tekton previously had with the Image PipelineResource.
```

Comment:
I am not sure I understand that sentence at all ? The Image PipelineResource did nothing on its own (you still had to use kaniko, buildah, … to generate the actual image).